### PR TITLE
refactor(intrinsics): migrate rag.py to new Adapter types (Epic #929 Phase 1 Wave 3)

### LINKED ISSUE #1137: refactor(intrinsics): rag.py whole-file migration to new Adapter types (Epic #929 Phase 1)
**Parent epic:** #929
**Design proposal:** PR #1080
**Phase:** 1 (Wave 3 — start after 1.A merges)
**Depends on:** 1.A
**Independent of:** 1.C, 1.D, 2.1 (all run in parallel in Wave 3)

---

## Problem

`mellea/stdlib/components/intrinsic/rag.py` contains six helpers in one file: `check_answerability`, `rewrite_question`, `clarify_query`, `find_citations`, `check_context_relevance`, `flag_hallucinated_content`. All construct via `IntrinsicAdapter(...)` (a deprecation shim after 1.A). Signatures do not consistently accept `model_options=`. None has output validation — schema drift from upstream `_RAG_REPO` weights silently changes behaviour.

Per-helper PRs would all touch the same file and serialise behind one another. One PR migrates the whole file.

## Agreed design

Follows the **per-helper-file migration template** from PR #1080:

1. **Migrate construction** — replace `IntrinsicAdapter(...)` with `Adapter(identity=..., io_contract=..., weights=...)`
2. **Normalise signature** — add `model_options: dict | None = None` as keyword argument to every helper
3. **Add output validation** — declare each helper's expected output contract; wire `io_contract.parse()` to raise `AdapterSchemaMismatchError` on contract-breaking deltas only (missing required field, type change). Forward-compatible additions (extra optional fields) do NOT raise.
4. **Update docs and examples** — docstrings must document declared output contract + new parameters; `docs/examples/intrinsics/` examples that call these helpers must be updated in the same PR

Six helpers, six declared output contracts. Implementer confirms each contract against current `_RAG_REPO` weights before writing the tests — document the verification in the PR description.

## Scope

- `mellea/stdlib/components/intrinsic/rag.py`
- Tests under `test/stdlib/components/intrinsic/test_rag.py`

## Out of scope

Other helper files (1.C, 1.D), backend changes (Phase 2), shim removal (4.1), refactoring beyond additive `model_options=`.

## Acceptance criteria

For each of the six helpers:

- [ ] Constructs via `Adapter(identity=..., io_contract=..., weights=...)` — no `IntrinsicAdapter(...)` calls in helper code
- [ ] Accepts `model_options: dict | None = None`
- [ ] Has a declared output contract documented in its docstring
- [ ] Forward-compat: synthetic output with an extra optional field does NOT raise
- [ ] Contract-break: synthetic output missing a required field raises `AdapterSchemaMismatchError`
- [ ] Existing helper tests pass (behavioural neutrality)
- [ ] `docs/examples/intrinsics/` examples that call these helpers updated; examples pass `uv run pytest docs/examples/intrinsics/` (or skipped with correct marker if backend-gated)
- [ ] `ruff format`, `ruff check`, `mypy` clean

## Test plan

Per helper (six pairs of tests):
- `test_<helper>_contract_enforced` — feed synthetic output missing a required field; assert `AdapterSchemaMismatchError`
- `test_<helper>_forward_compat` — feed output with an extra optional field; assert it does NOT raise

Existing happy-path tests pass unchanged.

> Note on `flag_hallucinated_content`: return type is boolean-like. Document expected output type explicitly in docstring; coercion responsibility stays with caller.

## Breaking changes

None at signature level (additive `model_options=`). Callers that previously swallowed `KeyError` on schema mismatch will now see `AdapterSchemaMismatchError` — surface in changelog.

## References

- PR #1080 §13
- Per-helper-file migration template in PR #1080 / `docs/dev/proposals/929-issue-plan.md`
- Jake req 4
- Before starting: `gh issue list --search "rag" --state open` — check for any open issues against specific RAG helpers and reference them in PR

### PR DESCRIPTION
**Pull request by planetf1**
## Summary

This is **Epic #929 Phase 1 Wave 3** — whole-file migration of `mellea/stdlib/components/intrinsic/rag.py` to the new `Adapter(identity, io_contract, weights)` dataclass system introduced in Phase 1.A (PR #1269, merged 2026-06-23).

Epic #929 is a multi-phase effort to fix Adapter Function Lifecycle & Consistency in Mellea. The phases are:
- **Phase 0 (#1134)** — scaffolding: `Adapter` dataclass, `IOContract` ABC, `Identity`, `WeightsBinding` types
- **Phase 1.A (#1136)** — `call_intrinsic` refactor + `_ShimIOContract` shims; merged ✅
- **Phase 1 Wave 3 (this PR, #1137)** — full IOContract migration for all 6 RAG adapter functions
- **Phase 2 (#1138)** — `build_prompt` implementation (currently stubs raising `NotImplementedError`)

### What this PR changes

**`rag.py`** — all six RAG adapter functions fully migrated:

- Two new `IOContract` subclasses: `_DictContract` (single-dict output) and `_ListContract` (list-of-dicts output).
- `_ListContract.parse()` wraps the validated list in `{"items": [...]}` to satisfy the `dict[str, object]` return type of `IOContract.parse()`; helper functions unwrap it before returning — public return types are unchanged.
- `parse()` raises `AdapterSchemaMismatchError` when a required field is missing; raises `ValueError` when the JSON output is the wrong type entirely (not a dict / not a list). This replaces the previous silent `KeyError`; **callers that catch `KeyError` should update to `AdapterSchemaMismatchError`**.
- Extra fields in the model output do NOT raise (forward-compatible).
- `parse()` propagates `ValueError` / `json.JSONDecodeError` from `json.loads` when the model output is not valid JSON.
- Six module-level `Adapter` constants, one per helper, each with `Identity`, `_*Contract`, and `LocalFileBinding()`.
- `model_options: dict | None = None` added as a keyword-only argument to all six helpers (previously absent).
- Prose updated to agreed Granite Switch glossary (issue #1192): "intrinsic" → "adapter function" in `backend:` parameter descriptions.
- All six call sites use `cast(str, ...)` / `cast(list[dict], ...)` rather than `# type: ignore[return-value]`.

**`_util.py`** — `call_intrinsic` extended with `io_contract: IOContract | None = None` (keyword-only, defaulted `None`). When provided, replaces the raw `json.loads` call with `io_contract.parse()` for combined parse+validate. All existing callers (`core.py`, `guardian.py`) pass no `io_contract` and are unaffected. The function now carries an explicit `-> dict[str, object]` return annotation.

**`test_rag_contracts.py`** (new) — 18 pure unit tests (no GPU, no backend):
- Two per helper: `contract_enforced` (missing required field → `AdapterSchemaMismatchError`) and `forward_compat` (extra field → no raise)
- Two empty-list edge cases for the list-shaped contracts (`find_citations`, `flag_hallucinated_content`)
- Four type-mismatch cases: `ValueError` when JSON is not a dict, not a list, or contains a non-dict list element

### Output contract verification

Each contract was verified against the current `granitelib-rag-r1.0` weights README before writing the tests:

| Helper | Required output keys |
|--------|---------------------|
| `check_answerability` | `answerability` |
| `rewrite_question` | `rewritten_question` |
| `clarify_query` | `clarification` |
| `find_citations` | `response_begin`, `response_end`, `response_text`, `citation_doc_id`, `citation_begin`, `citation_end`, `citation_text` |
| `check_context_relevance` | `context_relevance` |
| `flag_hallucinated_content` | `response_begin`, `response_end`, `response_text`, `faithfulness`, `explanation` |

### Examples review

All six `docs/examples/intrinsics/` examples that call these helpers (`answerability.py`, `citations.py`, `context_relevance.py`, `hallucination_detection.py`, `query_clarification.py`, `query_rewrite.py`) were reviewed. All existing calls pass the migrated parameters positionally or by name, and none rely on a `model_options` position — the keyword-only addition is non-breaking. All examples are gated with `# pytest: huggingface, e2e` (or have no pytest marker and are silently skipped) and therefore do not run in unit CI.

### Relationship to other open PRs

- **#1320** (`requirement_check_to_bool`, `core.py` — #1138): Wave 3 parallel sibling; touches different file, no conflict.
- **#1323** (`guardian.py` — #1139): Wave 3 parallel sibling; touches different file, no conflict.
- **#1322** (`docs/examples` terminology sweep — #1279): touches `docs/examples/intrinsics/README.md` and Python docstring files, but NOT `rag.py` or `_util.py`. No conflict.
- **#1284** (`.parsed` property / `format=` overloads — #1273/#1274): unrelated to this file.

Ordering: this PR has no blocking relationship to #1320 or #1323 — all three Wave 3 PRs are independent and can merge in any order.

### Related issues

- Closes #1137
- Part of Epic #929
- Prerequisite: #1136 / PR #1269 (merged — `Adapter`, `IOContract`, `LocalFileBinding` API)
- Next wave: #1138 (core.py, PR #1320), #1139 (guardian.py, PR #1323)
- Related open issues noted but out of scope: #1301 (context_relevance JSONDecodeError — separate bug), #937 (empty-context crash — separate bug), #1316 (groundedness nested citations — different helper)

### Test plan

- [x] `uv run pytest test/stdlib/components/intrinsic/test_rag_contracts.py -v` — 18/18 pass
- [x] `uv run pytest test/stdlib/components/intrinsic/ -m "not qualitative"` — all pass
- [x] `uv run pytest test/ -m "not qualitative"` (full local suite) — 1837 passed, 0 failures (1 pre-existing error: #1344)
- [x] `uv run ruff format . && uv run ruff check .` — clean
- [x] Docstring quality gate: 0 issues
- [x] `docs/examples/intrinsics/` examples reviewed — backward-compatible, no changes required
- [x] GPU integration tests — BlueVela LSF (job 1691399, single A100): 274 passed, 0 failures
- [x] CI (`code-checks` 3.11/3.12/3.13) — all pass (3.13 rerun needed: `httpx.ReadTimeout` flake, tracked #1318)

### CODE REVIEW

File: mellea/stdlib/components/intrinsic/rag.py [hunk 1]
```diff
@@ -31,58 +221,94 @@ def check_answerability(
...
         "answerability",
         context.add(
             Message("user", question, documents=_coerce_to_documents(documents))
         ),
         backend,
+        io_contract=_ANSWERABILITY_ADAPTER.io_contract,
+        model_options=model_options,
     )
-    return result_json["answerability"]
+    return result["answerability"]  # type: ignore[return-value]
```
Comment (jakelorocco) [inline mellea/stdlib/components/intrinsic/rag.py]: Should call_intrinsic be changed to allow for the correct return type here?

File: mellea/stdlib/components/intrinsic/rag.py [hunk 2]
```diff
@@ -1,26 +1,216 @@
...
+
+        Raises:
+            ValueError: When *raw* is not valid JSON.
+            AdapterSchemaMismatchError: When a required key is absent.
+        """
+        data = json.loads(raw)
+        if not isinstance(data, dict):
+            raise AdapterSchemaMismatchError(
+                self._name, frozenset(), self._required_keys
+            )
```
Comment (jakelorocco) [inline mellea/stdlib/components/intrinsic/rag.py:64]: I think this is misleading. This should be a type mismatch rather than a missing keys issue?

File: mellea/stdlib/components/intrinsic/rag.py (hunk 1 above)
Comment (planetf1) [reply mellea/stdlib/components/intrinsic/rag.py]: `_DictContract` and `_ListContract` now raise `ValueError` when the parsed JSON is the wrong shape entirely (not a dict / not a list), with `AdapterSchemaMismatchError` reserved for the missing-key case where its `observed_keys`/`expected_keys` vocabulary actually means something. Used `ValueError` rather than a new exception class — wrong-type output fits naturally in the same bucket as malformed JSON, and avoids adding public surface to `_core.py` mid-migration. Added four unit tests covering the new paths.

File: mellea/stdlib/components/intrinsic/rag.py (hunk 2 above)
Comment (planetf1) [reply mellea/stdlib/components/intrinsic/rag.py:64]: Added `-> dict[str, object]` to `call_intrinsic` and replaced the six `# type: ignore[return-value]` sites with `cast(str, ...)` / `cast(list[dict], ...)`. A cast at least names what the adapter contract guarantees about that key, rather than silencing whatever mypy finds on the line. Still unchecked at runtime — the guarantee comes from the contract, not the type system.

**Comment by planetf1** [comment]
**HPC results (job `1685548`, rebased branch `fbb206b7` on `upstream/main`):**

Intrinsic tests — `test/stdlib/components/intrinsic/`, `test/backends/test_adapters/`, `test/formatters/granite/test_intrinsics_formatters.py`, `docs/examples/intrinsics/`:

| | |
|---|---|
| Passed | 273 |
| Skipped | 2 |
| xfailed | 2 |
| Failed | **0** |

All green on CUDA (A100, BlueVela) and locally (MPS, Apple Silicon).

**Comment by planetf1** [comment]
## Testing

**Local (macOS/MPS)** — full non-qualitative suite:
```
uv run pytest test/ -m "not qualitative" -x --timeout=120 -q
```
Result: 1837 passed, 10 skipped, 1 error (pre-existing #1344 — MPS/docling float64), 0 failures.

**BlueVela (LSF, GPU)** — intrinsics scope (job 1691399):
```
uv run pytest test/stdlib/components/intrinsic/ test/backends/test_adapters/ \
    test/formatters/granite/test_intrinsics_formatters.py docs/examples/intrinsics/ -v
```
Result: 274 passed, 2 skipped, 2 xfailed, 0 failures. Run time: ~11 min on a single A100.

**Comment by planetf1** [comment]
The `code-checks / quality (3.13)` failure is a pre-existing flaky test — `test_ollama_generate_from_raw_metrics_integration` hit an `httpx.ReadTimeout` on a loaded CI runner, tracked in #1318. Restarting that job.

**Review by jakelorocco — APPROVED** [review]
